source/grid/grid_generator_pipe_junction.cc

```diff
@@ -0,0 +1,436 @@
+          Physics::VectorRelations::angle(directions[2], skeleton_unit[p]);
+        const vector rotation_axis = [&]() {
+          const vector rotation_axis =
+            cross_product_3d(directions[2], skeleton_unit[p]);
+          const double norm = rotation_axis.norm();
+          if (norm < tolerance)
+            return directions[1];
+          else
+            return rotation_axis / norm;
+        }();
+        const Tensor<2, spacedim, double> rotation_matrix =
+          Physics::Transformations::Rotations::rotation_matrix_3d(
+            rotation_axis, rotation_angle);
+        GridTools::transform(
+          [&](const Point<spacedim> &pt) { return rotation_matrix * pt; },
+          pipe);
+
+        //
+        // Step 4: rotate laterally to align pipe segments
+        //
+        // On the unit cylinder, we find that the edge on which all pipe
+        // segments meet is parallel to the x-axis. After the transformation to
+        // the pipe segment, we notice that this statement still holds for the
+        // projection of this edge onto the xy-plane, which corresponds to the
+        // cross section of the opening.
+        //
+        // With the latest rotation however, this is no longer the case. We
+        // rotate the unit vector in x-direction in the same fashion, which
+        // gives us the current direction of the projected edge.
+        const vector Rx = rotation_matrix * directions[0];
+
+        // To determine how far we need to rotate, we also need to project the
+        // polar axis of the bifurcation ball joint into the same plane.
+        const vector normal_projected_on_opening =
+          normal - (normal * skeleton_unit[p]) * skeleton_unit[p];
+
+        // Both the projected normal and Rx must be in the opening plane.
+        Assert(std::abs(skeleton_unit[p] * normal_projected_on_opening) <
+                 tolerance,
+               ExcInternalError());
+        Assert(std::abs(skeleton_unit[p] * Rx) < tolerance, ExcInternalError());
+
+        // Now we laterally rotate the pipe segment around its own symmetry axis
+        // that the edge matches the polar axis.
+        const double lateral_angle =
+          Physics::VectorRelations::signed_angle(Rx,
+                                                 normal_projected_on_opening,
+                                                 /*axis=*/skeleton_unit[p]);
+        GridTools::rotate(skeleton_unit[p], lateral_angle, pipe);
+
+        //
+        // Step 5: shift to final position
+        //
+        GridTools::shift(openings[p].first, pipe);
+
+        GridGenerator::merge_triangulations(
+          pipe, tria, tria, tolerance_length, /*copy_manifold_ids=*/true);
+      }
+
+    // Since GridGenerator::merge_triangulations() does not copy boundary IDs,
+    // we need to set them after the final geometry is created. Luckily,
+    // boundary IDs match with manifold IDs, so we simply translate them.
```

Comment:
```suggestion
    // boundary IDs match with material IDs, so we simply translate them.
```
As far as I understand this.

Reply:
Correct. We use manifold IDs to identify the openings.
```suggestion
    // boundary IDs match with material IDs, so we simply translate them with
    // the help of manifold IDs to identify openings.
```